float32 x_eff
float32 y_eff
